third_party/nvidia/lib/NVGPUToLLVM/NVGPUToLLVMPass.cpp

```diff
@@ -83,242 +82,186 @@ Type getTypeFromConstraint(char constraint, mlir::PatternRewriter &rewriter) {
+class NVGPUOpGenericPattern : public OpRewritePattern<SourceOp> {
 public:
-  explicit NVGPUOpGenericPattern(mlir::MLIRContext *context, std::string ptxAsm,
-                                 std::vector<std::string> outputConstraints,
-                                 std::vector<std::string> inputConstraints)
-      : NVGPUOpPatternBase<SourceOp, NVGPUOpGenericPattern<SourceOp>>(context),
-        ptxAsm(ptxAsm), outputConstraints(outputConstraints),
+  explicit NVGPUOpGenericPattern(MLIRContext *context, std::string ptxAsm,
+                                 Constraints outputConstraints,
+                                 Constraints inputConstraints)
+      : OpRewritePattern<SourceOp>(context), ptxAsm(std::move(ptxAsm)),
+        outputConstraints(outputConstraints),
         inputConstraints(inputConstraints) {}
 
-  std::vector<std::string> getOutputConstraints(SourceOp op) const {
-    return outputConstraints;
-  }
-  OperandsAndConstraints getOperandsAndConstraints(SourceOp op) const {
+  LogicalResult matchAndRewrite(SourceOp op,
+                                PatternRewriter &rewriter) const override {
     OperandsAndConstraints operandsAndConstraints;
     for (unsigned i = 0; i < inputConstraints.size(); i++) {
       operandsAndConstraints.push_back(
           {op->getOperand(i), inputConstraints[i]});
     }
-    return operandsAndConstraints;
+    return rewriteAsPtxAsm(op, rewriter, ptxAsm, operandsAndConstraints,
+                           outputConstraints);
   }
-  std::string getPtxAsm(SourceOp op) const { return ptxAsm; }
 
 private:
   std::string ptxAsm;
-  std::vector<std::string> outputConstraints;
-  std::vector<std::string> inputConstraints;
+  Constraints outputConstraints;
+  Constraints inputConstraints;
 };
 
 class FenceAsyncSharedOpPattern
-    : public NVGPUOpPatternBase<ttn::FenceAsyncSharedOp,
-                                FenceAsyncSharedOpPattern> {
+    : public OpRewritePattern<ttn::FenceAsyncSharedOp> {
 public:
-  using Base =
-      NVGPUOpPatternBase<ttn::FenceAsyncSharedOp, FenceAsyncSharedOpPattern>;
-  using Base::Base;
-
-  std::string getPtxAsm(ttn::FenceAsyncSharedOp op) const {
-    auto bCluster = op.getBCluster();
-    if (bCluster)
-      return "fence.proxy.async.shared::cluster;";
-    else
-      return "fence.proxy.async.shared::cta;";
+  using OpRewritePattern<ttn::FenceAsyncSharedOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(ttn::FenceAsyncSharedOp op,
+                                PatternRewriter &rewriter) const override {
+    std::string ptxAsm = op.getBCluster() ? "fence.proxy.async.shared::cluster;"
+                                          : "fence.proxy.async.shared::cta;";
+    return rewriteAsPtxAsm(op, rewriter, std::move(ptxAsm));
   }
 };
 
 class ClusterArriveOpPattern
-    : public NVGPUOpPatternBase<ttn::ClusterArriveOp, ClusterArriveOpPattern> {
+    : public OpRewritePattern<ttn::FenceAsyncSharedOp> {
```

Comment:
why did this become `FenceAsyncSharedOp`?

Reply:
Sorry, probably a copy-paste mistake. Fixed.

Reply:
These patterns don't seem to be covered by tests. I can follow up with that.

Reply:
indeed, yes if you can help here that would be great